apiserver/listener.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package apiserver
+
+import (
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/juju/utils/clock"
+)
+
+func newThrottlingListener(inner net.Listener, minPause, maxPause time.Duration, clk clock.Clock) net.Listener {
+	rand.Seed(time.Now().UTC().UnixNano())
+	if clk == nil {
+		clk = clock.WallClock
+	}
+	return &throttlingListener{
+		Listener:        inner,
+		maxPause:        maxPause,
+		minPause:        minPause,
+		clk:             clk,
+		connAcceptTimes: make([]*time.Time, 500),
+	}
+}
+
+type throttlingListener struct {
+	sync.Mutex
+	net.Listener
+	connAcceptTimes []*time.Time
+	nextSlot        int
+
+	minPause time.Duration
+	maxPause time.Duration
+	clk      clock.Clock
+}
+
+// connRateMetric returns an int value based on the rate of new connections.
+func (l *throttlingListener) connRateMetric() int {
+	l.Lock()
```

Comment:
The averaging over possible stale connections issue is a fair point - I could put a bound on how far back in time to look. This would make having a last slot pointer irrelevant, as we'd walk back until we wrapped or exceeded the max lookback time